File: Java/546.java
__________________________________________________________________________________________________
sample 3 ms submission
class Solution {
    int[][] dp;
    int[] boxes;
    public int removeBoxes(int[] boxes) {
        int len = boxes.length;
        dp = new int[len][len];

        Map<Integer, Integer> last = new HashMap<>();
        for (int i = boxes.length - 1; i >= 0; --i) {
            int color = boxes[i];
            boxes[i] = last.getOrDefault(color, len);
            last.put(color, i);
        }

        this.boxes = boxes;
        return dpc(0, len - 1);
    }

    int dpc(int i, int j) {
        if (i == j) return 1;
        if (i > j) return 0;
        int r = dp[i][j];
        if (r > 0) return r;

        return dp[i][j] = dfs(j, i, boxes[i], 1, 0);
    }

    int dfs(int j, int x, int y, int count, int part) {
        if (y > j) {
            return count*count + part + (x < j ? dpc(x+1, j) : 0);
        }

        if (x+1 == y) {
            return dfs(j, y, boxes[y], count+1, part);
        }

        return Math.max(dfs(j, x, boxes[y], count, part),
            dfs(j, y, boxes[y], count + 1, part + dpc(x+1, y-1)));
    }
}
__________________________________________________________________________________________________
sample 49796 kb submission
class Solution {
    public int removeBoxes(int[] boxes) {

        int n = boxes.length;
        int[][][] dp = new int[n][n][n];

        for (int j = 0; j < n; j++) 
            for (int k = 0; k <= j; k++) 
                dp[j][j][k] = (k + 1) * (k + 1);



        for (int l = 1; l < n; l++) 
            for (int j = l; j < n; j++) {
                int i = j - l;

                for (int k = 0; k <= i; k++) {
                    int res = (k + 1) * (k + 1) + dp[i + 1][j][0];

                    for (int m = i + 1; m <= j; m++) 
                        if (boxes[m] == boxes[i]) 
                            res = Math.max(res, dp[i + 1][m - 1][0] + dp[m][j][k + 1]);

                    dp[i][j][k] = res;
                }
            }


        return (n == 0 ? 0 : dp[0][n - 1][0]);
    }
}
__________________________________________________________________________________________________

File: Python3/546.py
__________________________________________________________________________________________________
sample 72 ms submission
from  collections import defaultdict
class Solution:
    def removeBoxes(self, boxes):
        if not boxes:
            return 0
        tmp = []
        color, count = boxes[0], 1
        for i in range(1, len(boxes)):
            if color != boxes[i]:
                tmp.append([color, count])
                color, count = boxes[i], 1
            else:
                count += 1
        tmp.append([color, count])

        lst_index = defaultdict(list)
        dt_index = defaultdict(dict)
        for i in range(len(tmp)):
            lst_index[tmp[i][0]].append(i)
            dt_index[tmp[i][0]][i] = len(lst_index[tmp[i][0]]) - 1

        memo = {}
        def helper(i, j, k):
            if i > j:
                return 0
            if i == j:
                return k * k
            if (i, j, k) in memo:
                return memo[i, j, k]
            res = k * k + helper(i, j - 1, tmp[j - 1][1])

            ttt = lst_index[tmp[j][0]]
            iii = dt_index[tmp[j][0]][j] - 1
            while iii >= 0 and ttt[iii] >= i:
                idx = ttt[iii]
                res = max(res, helper(i, idx, tmp[idx][1] + k) + helper(idx + 1, j - 1, tmp[j - 1][1]))
                iii -= 1
            memo[i, j, k] = res
            return res

        return helper(0, len(tmp) - 1, tmp[-1][1])
__________________________________________________________________________________________________
sample 76 ms submission
from  collections import defaultdict


class Solution:
    def removeBoxes(self, boxes):
        if not boxes:
            return 0
        tmp = []
        color, count = boxes[0], 1
        for i in range(1, len(boxes)):
            if color != boxes[i]:
                tmp.append([color, count])
                color, count = boxes[i], 1
            else:
                count += 1
        tmp.append([color, count])

        lst_index = defaultdict(list)
        dt_index = defaultdict(dict)
        for i in range(len(tmp)):
            lst_index[tmp[i][0]].append(i)
            dt_index[tmp[i][0]][i] = len(lst_index[tmp[i][0]]) - 1

        memo = {}
        def helper(i, j, k):
            if i > j:
                return 0
            if i == j:
                return k * k
            if (i, j, k) in memo:
                return memo[i, j, k]
            res = k * k + helper(i, j - 1, tmp[j - 1][1])

            ttt = lst_index[tmp[j][0]]
            iii = dt_index[tmp[j][0]][j] - 1
            while iii >= 0 and ttt[iii] >= i:
                idx = ttt[iii]
                res = max(res, helper(i, idx, tmp[idx][1] + k) + helper(idx + 1, j - 1, tmp[j - 1][1]))
                iii -= 1
            memo[i, j, k] = res
            return res

        return helper(0, len(tmp) - 1, tmp[-1][1])
__________________________________________________________________________________________________
